android/kotlin/v2/AmpliApp/app/src/main/java/com/amplitude/ampli/Ampli.kt

```diff
@@ -0,0 +1,822 @@
+     */
+    constructor(
+        optionalArrayNumber: Array<Double>? = null,
+        optionalArrayString: Array<String>? = null,
+        optionalBoolean: Boolean? = null,
+        optionalNumber: Double? = null,
+        optionalString: String? = null
+    ) : this() {
+        this.eventType = "EventWithOptionalProperties"
+        this.eventProperties =
+            mapOf(
+                *(if (optionalArrayNumber != null) arrayOf("optionalArrayNumber" to optionalArrayNumber) else arrayOf()),
+                *(if (optionalArrayString != null) arrayOf("optionalArrayString" to optionalArrayString) else arrayOf()),
+                *(if (optionalBoolean != null) arrayOf("optionalBoolean" to optionalBoolean) else arrayOf()),
+                *(if (optionalNumber != null) arrayOf("optionalNumber" to optionalNumber) else arrayOf()),
+                *(if (optionalString != null) arrayOf("optionalString" to optionalString) else arrayOf())
+            ).toMutableMap()
+    }
+  }
+
+val ampli = Ampli()
+
+open class Ampli {
+    companion object {
+        val API_KEY: Map<Environment, String> = mapOf(
+            Environment.DEVELOPMENT to "",
+            Environment.PRODUCTION to ""
+        )
+
+		private val observePlan: Plan = Plan("main", "kotlin-ampli-v2", "0", "79154a50-f057-4db5-9755-775e4e9f05e6")
+    }
+
+    enum class Environment {
+        DEVELOPMENT,
+        PRODUCTION
+    }
+
+    private var disabled: Boolean = false
+
+    @android.annotation.SuppressLint("StaticFieldLeak")
+    private var _client: Amplitude? = null
+
+    open fun load(appContext: android.content.Context, options: LoadOptions? = null) {
+        this.disabled = options?.disabled ?: false
+        if (this._client != null) {
+            System.err.println("Warning: Ampli is already initialized. ampli.load() should be called once at application start up.")
+            return
+        }
+        val env = options?.environment ?: Environment.DEVELOPMENT
+        val apiKey = options?.client?.apiKey ?: API_KEY[env]
+        val plan = options?.client?.plan ?: observePlan
+        when {
+            options?.client?.instance != null -> {
+                this._client = options.client.instance
+            }
+            apiKey != null && apiKey != "" -> {
+                this._client = Amplitude(
+                    Configuration(
+                        apiKey = apiKey,
+                        plan = plan,
+                        context = appContext.applicationContext
+                    )
+                )
+            }
+            else -> {
+                System.err.println("ampli.load() requires 'environment', 'client.apiKey', or 'client.instance'")
+                return
+            }
+        }
+    }
+
+    open fun track(event: BaseEvent, options: EventOptions? = null) {
+		if (!isInitializedAndEnabled()) {
+            return
+        }
+        this._client?.track(event.eventType, this.getEventPropertiesMap(event), options)
```

Comment:
could we just call track(event, options)?

Reply:
The function of track has 2 ways
1. `fun track(event: BaseEvent, callback: EventCallBack? = null)`
2. `fun track(eventType: String, eventProperties: Map<String, Any?>? = null, options: EventOptions? = null)`
There is no `track(event: BaseEvent, options: EventOptions? = null)`. Feels we can merge the event with options first and then call track(event) like
` open fun track(event: BaseEvent, options: EventOptions? = null) {
        if (!isInitializedAndEnabled()) {
            return
        }
        event.mergeEventOptions(options)
        this._client?.track(event)
    }`

But the options in mergeEventOptions are required to be com.amplitude.core.EventOptions no com.amplitude.android.EventOptions. So maybe we need to import com.amplitude.android.EventOptions in BaseEvent class in Kotlin? 
<img width="721" alt="Screen Shot 2022-05-27 at 10 46 54 PM" src="https://user-images.githubusercontent.com/13028329/170812001-094d8cdc-eb59-4ef5-a58e-f7275627bf12.png">